src/python/dxpy/sugar/objects.py

```diff
@@ -0,0 +1,422 @@
+                )
+            else:
+                LOG.info("Project %s exists in region %s", project.name, project_region)
+        except dxpy.exceptions.ResourceNotFound:
+            LOG.info("Project with ID %s not found", id_or_name)
+            project = None
+
+    if project:
+        if exists is False:
+            raise dxpy.AppError(
+                "Project {} exists but was expected not to exist".format(id_or_name)
+            )
+        else:
+            return project
+
+    LOG.info("Searching for project with name %s", id_or_name)
+    find_args = dict(kwargs)
+    if region:
+        find_args["region"] = region
+    if level:
+        find_args["level"] = level
+    found = list(dxpy.find_projects(id_or_name, **find_args))
+
+    if len(found) == 0 and exists is not True and create:
+        LOG.info(
+            "Creating project %s in %s region",
+            id_or_name, region or "<default>"
+        )
+        create_args = dict(kwargs)
+        create_args["name"] = id_or_name
+        if region:
+            create_args["region"] = region
+        project_id = dxpy.api.project_new(create_args)["id"]
+        return dxpy.DXProject(project_id)
+
+    return _create_one_handler(found, "project", id_or_name, exists=exists)
+
+
+def ensure_folder(folder, project=None, exists=None, create=False):
+    """Checks that the given project contains (or does not contain) the given folder.
+    The folder is created if it does not exist and `create is True`.
+
+    Args:
+        folder: The folder to validate.
+        project: Project ID, name, or DXProject object. The project in which to look
+            for the folder. Defaults to the current project.
+        exists: Assert whether the folder exists (True) or does not exist (False).
+        create: Whether to create the folder if it does not exist.
+
+    Returns:
+        A listing of the objects in the folder.
+
+    Raises:
+        dxpy.AppError if the folder does exist and `create is True` or does not exist
+            and `create is False`.
+
+    """
+    folder = folder.strip()
+    project = get_project(project)
+
+    LOG.info(
+        "Checking whether folder %s exists in project %s", folder, project.get_id()
+    )
+
+    try:
+        ls = project.list_folder(folder=folder)
+    except dxpy.exceptions.ResourceNotFound:
+        ls = None
+
+    if ls is None:
+        if exists is True:
+            LOG.error(
+                "Folder %s not found in project %s", folder, project.get_id(),
+                exc_info=True
+            )
+            raise dxpy.AppError(
+                "Folder {} does not exist in project {}".format(
+                    folder, project.get_id()
+                )
+            )
+        elif create:
+            LOG.info("Creating new folder %s in project %s", folder, project.get_id())
+            project.new_folder(folder, parents=True)
+            return []
+    elif exists is False:
+        raise dxpy.AppError(
```

Comment:
AppErrors are specific to execution environments (job environments). Here a `dxpy.DXSearchError` is preferred

```suggestion
        raise dxpy.DXSearchError(
```